File: src/utils/binaryDetection.ts
// Binary file extensions - keep in sync with Rust binary_detection.rs
const IMAGE_EXTENSIONS = [
  'png', 'jpg', 'jpeg', 'gif', 'bmp', 'tiff', 'tif', 'webp', 'ico', 'svg',
] as const

const OTHER_BINARY_EXTENSIONS = [
  // Audio files
  'mp3', 'wav', 'flac', 'aac', 'ogg', 'm4a', 'wma',
  // Video files
  'mp4', 'avi', 'mkv', 'mov', 'wmv', 'flv', 'webm', 'm4v',
  // Archive files
  'zip', 'rar', '7z', 'tar', 'gz', 'bz2', 'xz', 'lz4', 'lzma',
  // Executable files
  'exe', 'dll', 'so', 'dylib', 'bin', 'app', 'deb', 'rpm', 'dmg', 'pkg',
  // Office files
  'doc', 'docx', 'xls', 'xlsx', 'ppt', 'pptx', 'pdf', 'odt', 'ods', 'odp',
  // Database files
  'db', 'sqlite', 'sqlite3', 'mdb', 'accdb',
  // Font files
  'ttf', 'otf', 'woff', 'woff2', 'eot',
  // Other binary formats
  'pyc', 'class', 'jar', 'war', 'ear', 'o', 'obj', 'lib', 'a'
] as const

const BINARY_EXTENSIONS = [
  ...IMAGE_EXTENSIONS,
  ...OTHER_BINARY_EXTENSIONS,
] as const

// Create a Set for efficient lookups
const IMAGE_EXTENSIONS_SET = new Set(IMAGE_EXTENSIONS)
const BINARY_EXTENSIONS_SET = new Set(BINARY_EXTENSIONS)

/**
 * Check if a file is likely binary based on its extension
 * @param filePath - The file path to check
 * @returns true if the file extension indicates a binary file
 */
export function isBinaryFileByExtension(filePath: string): boolean {
  if (!filePath) return false
  
  const ext = filePath.split('.').pop()?.toLowerCase()
  if (!ext) return false
  
  return BINARY_EXTENSIONS_SET.has(ext as typeof BINARY_EXTENSIONS[number])
}

/**
 * Get all supported binary file extensions
 * @returns Array of binary file extensions
 */
export function getBinaryExtensions(): readonly string[] {
  return BINARY_EXTENSIONS
}

export function getImageExtensions(): readonly string[] {
  return IMAGE_EXTENSIONS
}

/**
 * Check if a file extension is binary
 * @param extension - File extension (without dot)
 * @returns true if extension is for a binary file type
 */
export function isBinaryExtension(extension: string): boolean {
  return BINARY_EXTENSIONS_SET.has(extension.toLowerCase() as typeof BINARY_EXTENSIONS[number])
}

export function isImageFileByExtension(filePath: string): boolean {
  if (!filePath) return false

  const ext = filePath.split('.').pop()?.toLowerCase()
  if (!ext) return false

  return IMAGE_EXTENSIONS_SET.has(ext as typeof IMAGE_EXTENSIONS[number])
}

File: src-tauri/src/binary_detection.rs
static IMAGE_EXTENSIONS: &[&str] = &[
    "png", "jpg", "jpeg", "gif", "bmp", "tiff", "tif", "webp", "ico", "svg",
];

static BINARY_EXTENSIONS: &[&str] = &[
    // Image files
    "png", "jpg", "jpeg", "gif", "bmp", "tiff", "tif", "webp", "ico", "svg",
    // Audio files
    "mp3", "wav", "flac", "aac", "ogg", "m4a", "wma", // Video files
    "mp4", "avi", "mkv", "mov", "wmv", "flv", "webm", "m4v", // Archive files
    "zip", "rar", "7z", "tar", "gz", "bz2", "xz", "lz4", "lzma", // Executable files
    "exe", "dll", "so", "dylib", "bin", "app", "deb", "rpm", "dmg", "pkg",
    // Office files
    "doc", "docx", "xls", "xlsx", "ppt", "pptx", "pdf", "odt", "ods", "odp",
    // Database files
    "db", "sqlite", "sqlite3", "mdb", "accdb", // Font files
    "ttf", "otf", "woff", "woff2", "eot", // Other binary formats
    "pyc", "class", "jar", "war", "ear", "o", "obj", "lib", "a",
];

pub fn is_binary_file_by_extension(file_path: &str) -> bool {
    if file_path.is_empty() {
        return false;
    }

    let ext = match file_path.split('.').next_back() {
        Some(e) => e.to_lowercase(),
        None => return false,
    };

    BINARY_EXTENSIONS.iter().any(|candidate| *candidate == ext)
}

pub fn is_image_file_by_extension(file_path: &str) -> bool {
    if file_path.is_empty() {
        return false;
    }

    let ext = match file_path.split('.').next_back() {
        Some(e) => e.to_lowercase(),
        None => return false,
    };

    IMAGE_EXTENSIONS.iter().any(|candidate| *candidate == ext)
}

pub fn image_mime_type(file_path: &str) -> Option<&'static str> {
    let ext = file_path.split('.').next_back()?.to_lowercase();
    match ext.as_str() {
        "png" => Some("image/png"),
        "jpg" | "jpeg" => Some("image/jpeg"),
        "gif" => Some("image/gif"),
        "bmp" => Some("image/bmp"),
        "tiff" | "tif" => Some("image/tiff"),
        "webp" => Some("image/webp"),
        "ico" => Some("image/x-icon"),
        "svg" => Some("image/svg+xml"),
        _ => None,
    }
}

pub fn is_likely_binary_content(bytes: &[u8]) -> bool {
    // Use Git's standard algorithm: check for null bytes in first 8000 bytes
    // This matches Git's buffer_is_binary() function
    let check_size = std::cmp::min(8000, bytes.len());
    let sample = &bytes[..check_size];

    // Check for null bytes (Git's standard binary detection)
    sample.contains(&0)
}

pub fn get_unsupported_reason(file_path: &str, content_bytes: Option<&[u8]>) -> Option<String> {
    if is_binary_file_by_extension(file_path) {
        return Some(format!(
            "Binary file type ({})",
            file_path.split('.').next_back().unwrap_or("unknown")
        ));
    }

    if let Some(bytes) = content_bytes {
        if bytes.len() > 10 * 1024 * 1024 {
            return Some("File is too large to diff (>10MB)".to_string());
        }

        if is_likely_binary_content(bytes) {
            return Some("File contains binary data".to_string());
        }
    }

    None
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_is_binary_file_by_extension() {
        // Test image files
        assert!(is_binary_file_by_extension("test.png"));
        assert!(is_binary_file_by_extension("image.jpg"));
        assert!(is_binary_file_by_extension("photo.JPEG"));

        // Test archive files
        assert!(is_binary_file_by_extension("archive.zip"));
        assert!(is_binary_file_by_extension("compressed.tar.gz"));

        // Test executable files
        assert!(is_binary_file_by_extension("program.exe"));
        assert!(is_binary_file_by_extension("library.dll"));
        assert!(is_binary_file_by_extension("lib.so"));

        // Test office files
        assert!(is_binary_file_by_extension("document.pdf"));
        assert!(is_binary_file_by_extension("spreadsheet.xlsx"));

        // Test text files (should return false)
        assert!(!is_binary_file_by_extension("code.rs"));
        assert!(!is_binary_file_by_extension("text.txt"));
        assert!(!is_binary_file_by_extension("config.json"));
        assert!(!is_binary_file_by_extension("style.css"));
        assert!(!is_binary_file_by_extension("script.js"));

        // Test edge cases
        assert!(!is_binary_file_by_extension(""));
        assert!(!is_binary_file_by_extension("no_extension"));
        assert!(!is_binary_file_by_extension("multiple.dots.txt"));
    }

    #[test]
    fn test_is_likely_binary_content() {
        // Test text content
        let text_content = b"Hello, world!\nThis is a text file.";
        assert!(!is_likely_binary_content(text_content));

        // Test binary content with null bytes
        let binary_content = b"Hello\x00World\x00Binary";
        assert!(is_likely_binary_content(binary_content));

        // Test empty content
        let empty_content = b"";
        assert!(!is_likely_binary_content(empty_content));

        // Test large text content
        let large_text = "A".repeat(10000).into_bytes();
        assert!(!is_likely_binary_content(&large_text));

        // Test content with null byte at the end (within check limit)
        let mut text_with_null = "Text content".to_string().into_bytes();
        text_with_null.push(0);
        assert!(is_likely_binary_content(&text_with_null));
    }

    #[test]
    fn test_get_unsupported_reason() {
        // Test binary file by extension
        let reason = get_unsupported_reason("test.png", None);
        assert!(reason.is_some());
        assert!(reason.unwrap().contains("Binary file type (png)"));

        // Test large file
        let large_content = vec![b'A'; 11 * 1024 * 1024]; // 11MB
        let reason = get_unsupported_reason("large.txt", Some(&large_content));
        assert!(reason.is_some());
        assert!(reason.unwrap().contains("too large"));

        // Test binary content
        let binary_content = b"Hello\x00World";
        let reason = get_unsupported_reason("test.txt", Some(binary_content));
        assert!(reason.is_some());
        assert!(reason.unwrap().contains("binary data"));

        // Test normal text file
        let text_content = b"Hello, world!";
        let reason = get_unsupported_reason("test.txt", Some(text_content));
        assert!(reason.is_none());
    }

    #[test]
    fn test_case_insensitive_extensions() {
        assert!(is_binary_file_by_extension("IMAGE.PNG"));
        assert!(is_binary_file_by_extension("Document.PDF"));
        assert!(is_binary_file_by_extension("Music.MP3"));
        assert!(is_binary_file_by_extension("Video.Mp4"));
    }

    #[test]
    fn test_is_image_file_by_extension() {
        assert!(is_image_file_by_extension("asset.png"));
        assert!(is_image_file_by_extension("asset.JPG"));
        assert!(is_image_file_by_extension("icons/logo.svg"));
        assert!(!is_image_file_by_extension("archive.zip"));
        assert!(!is_image_file_by_extension("document.pdf"));
        assert!(!is_image_file_by_extension("src/main.rs"));
    }

    #[test]
    fn test_image_mime_type() {
        assert_eq!(image_mime_type("asset.png"), Some("image/png"));
        assert_eq!(image_mime_type("photo.jpeg"), Some("image/jpeg"));
        assert_eq!(image_mime_type("photo.jpg"), Some("image/jpeg"));
        assert_eq!(image_mime_type("icon.svg"), Some("image/svg+xml"));
        assert_eq!(image_mime_type("graphic.webp"), Some("image/webp"));
        assert_eq!(image_mime_type("archive.zip"), None);
    }

    #[test]
    fn test_extension_consistency() {
        // Verify all extensions in the static array are detected correctly
        for ext in BINARY_EXTENSIONS.iter() {
            let test_file = format!("test.{}", ext);
            assert!(
                is_binary_file_by_extension(&test_file),
                "Extension {} should be detected as binary",
                ext
            );
        }
    }
}
